deepspeed/profiling/flops_profiler/profiler.py

```diff
@@ -0,0 +1,751 @@
+    active_elements_count = batch_size * int(prod(input_dims))
+    overall_conv_flops = conv_per_position_flops * active_elements_count
+
+    bias_flops = 0
+    if bias is not None:
+        bias_flops = out_channels * batch_size * int(prod(output_dims))
+
+    overall_flops = overall_conv_flops + bias_flops
+
+    return int(overall_flops)
+
+
+def batch_norm_flops_compute(
+    input,
+    running_mean,
+    running_var,
+    weight=None,
+    bias=None,
+    training=False,
+    momentum=0.1,
+    eps=1e-05,
+):
+    # assume affine is true
+    flops = 2 * torch.numel(input)
+    return flops
+
+
+def upsample_flops_compute(input,
+                           size=None,
+                           scale_factor=None,
+                           mode="nearest",
+                           align_corners=None):
+    if size is not None:
+        return int(prod(size))
+    assert scale_factor is not None
+    flops = torch.numel(input)
+    if len(scale_factor) == len(input):
+        flops * int(prod(scale_factor))
+    else:
+        flops * scale_factor**len(input)
+    return flops
+
+
+def softmax_flops_compute(input, dim=None, _stacklevel=3, dtype=None):
+    return torch.numel(input)
+
+
+def embedding_flops_compute(
+    input,
+    weight,
+    padding_idx=None,
+    max_norm=None,
+    norm_type=2.0,
+    scale_grad_by_freq=False,
+    sparse=False,
+):
+    return 0
+
+
+def dropout_flops_compute(input, p=0.5, training=True, inplace=False):
+    return 0
+
+
+def wrapFunc(func, funcFlopCompute):
+    oldFunc = func
+
+    def newFunc(*args, **kwds):
+        flops = funcFlopCompute(*args, **kwds)
+        name = "nn.functional." + func.__name__
+        module_flop_count.append((name, flops))
+        return oldFunc(*args, **kwds)
+
+    return newFunc
+
+
+# FC
+F.linear = wrapFunc(F.linear, linear_flops_compute)
+
+# convolutions
+F.conv1d = wrapFunc(F.conv1d, conv_flops_compute)
+F.conv2d = wrapFunc(F.conv2d, conv_flops_compute)
+F.conv3d = wrapFunc(F.conv3d, conv_flops_compute)
+
+# conv transposed
+F.conv_transpose1d = wrapFunc(F.conv_transpose1d, conv_trans_flops_compute)
+F.conv_transpose2d = wrapFunc(F.conv_transpose2d, conv_trans_flops_compute)
+F.conv_transpose3d = wrapFunc(F.conv_transpose3d, conv_trans_flops_compute)
+
+# activations
+F.relu = wrapFunc(F.relu, relu_flops_compute)
+F.prelu = wrapFunc(F.prelu, relu_flops_compute)
+F.elu = wrapFunc(F.elu, relu_flops_compute)
+F.leaky_relu = wrapFunc(F.leaky_relu, relu_flops_compute)
+F.relu6 = wrapFunc(F.relu6, relu_flops_compute)
+
+# BatchNorms
+F.batch_norm = wrapFunc(F.batch_norm, batch_norm_flops_compute)
+
+# poolings
+F.avg_pool1d = wrapFunc(F.avg_pool1d, pool_flops_compute)
+F.avg_pool2d = wrapFunc(F.avg_pool2d, pool_flops_compute)
+F.avg_pool3d = wrapFunc(F.avg_pool3d, pool_flops_compute)
+F.max_pool1d = wrapFunc(F.max_pool1d, pool_flops_compute)
+F.max_pool2d = wrapFunc(F.max_pool2d, pool_flops_compute)
```

Comment:
It would be cool to have a data structure that maps a list of functionals to the compute wrapper, and then at wrap time the engine code just traverses that data structure to do this work.

Reply:
Monkey patching does not work if the target function is referenced as an item in a loop.